cpp/src/text/replace.cu

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/strings/detail/utilities.hpp>
+#include <cudf/strings/string_view.cuh>
+#include <cudf/strings/strings_column_view.hpp>
+#include <cudf/utilities/error.hpp>
+#include <nvtext/detail/tokenize.hpp>
+#include <nvtext/tokenize.hpp>
+#include <strings/utilities.cuh>
+#include <text/utilities/tokenize_ops.cuh>
+
+namespace nvtext {
+namespace detail {
+namespace {
+
+/**
+ * @brief Functor to replace tokens in each string.
+ *
+ * This tokenizes a string using the given d_delimiter and replaces any tokens that match
+ * a string in d_targets with those from the d_repls column. Strings with no matching tokens
+ * are left unchanged.
+ *
+ * This should be called first to compute the size of each output string and then a second
+ * time to fill in the allocated output buffer for each string.
+ */
+struct replace_tokens_fn {
+  cudf::column_device_view const d_strings;  ///< strings to tokenize
+  cudf::column_device_view const d_targets;  ///< strings to replace
+  cudf::column_device_view const d_repls;    ///< replacement strings
+  cudf::string_view const d_delimiter;       ///< delimiter characters for tokenizing
+  const int32_t* d_offsets{};
+  char* d_chars{};
+
+  __device__ cudf::size_type operator()(cudf::size_type idx)
+  {
+    if (d_strings.is_null(idx)) return 0;
+    cudf::string_view const d_str = d_strings.element<cudf::string_view>(idx);
+    cudf::size_type nbytes        = d_str.size_bytes();
+    auto in_ptr                   = d_str.data();
+    char* out_ptr                 = d_chars ? d_chars + d_offsets[idx] : nullptr;
+    cudf::size_type last_pos      = 0;
```

Comment:
My personal preference is to either have all `auto`s or all explicit.
Change to:
```cpp
    auto const d_str = d_strings.element<cudf::string_view>(idx);
    auto nbytes      = d_str.size_bytes();
    auto in_ptr      = d_str.data();
    auto out_ptr     = d_chars ? d_chars + d_offsets[idx] : nullptr;
    auto last_pos    = cudf::size_type{0};
```

Reply:
My personal preference is to be able to know the type when I read the code unless the type is unutterable.

Reply:
`cudf::string_view::data()` returns a `const char*` so I'm fine changing the `auto`. Just seems odds to use a single `auto` amongst other explicit types.

Reply:
I don't think we should have a rule that you can't mix auto and explicit types. I think that's silly.  What if you have a bunch of `int`s and a single unutterable type? You shouldn't have to make the `int`s `auto` just to match the one auto.

Reply:
My bad - maybe I miscommunicated. I am definitely not saying we make this a "rule." I was just stating: 
> My personal preference ...

And to elaborate, I really like symmetry and alignment. There is value in symmetry. There even at times (i.e. when it comes to naming things) is meaning and semantics in symmetry (take for example `std::transform` vs `std::adjacent_difference`: the lack of symmetry in naming obfuscates the fact one algorithm is a specialisation of the other). That is why I prefer consistency. Consistency is a form of symmetry. And alignment (which is also a form of symmetry), makes things easier to read.

On top of this, there is still quite a bit of name type information in my version:
```cpp
auto const d_str = d_strings.element<cudf::string_view>(idx);    // still says string_view
auto nbytes      = d_str.size_bytes();                           // mostly likely is a size type
auto in_ptr      = d_str.data();                                 // know is a ptr
auto out_ptr     = d_chars ? d_chars + d_offsets[idx] : nullptr; // know is a ptr
auto last_pos    = cudf::size_type{0};                           // still says cudf::size_type
```
So line `1` and `5` you know the type still and the rest you still have a good idea. I know there is a trade-off. Was just highlighting my preference. I am fine not changing anything at the end of the day.